enhancements/aws-ipv6-dev-env.md

```diff
@@ -0,0 +1,163 @@
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2019-10-24
+last-updated: 2019-10-24
+status: implementable
+---
+
+# aws-ipv6-dev-env
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+While Kubernetes itself includes support for IPv6, the OKD distribution does
+not fully support it yet.  To enable developers across the many components that
+make up OKD, this proposal is to add support for enabling IPv6 for an AWS
+cluster.  This will also be used to create a CI job that can exercise IPv6
+support across OKD.
+
+## Motivation
+
+Supporting IPv6 across OKD will require collaboration and fixes across many
+components.  Many developers are already used to doing development against AWS
+clusters, so supporting AWS as an IPv6 development environment should make it
+easier for people to jump in to help debug and fix IPv6 issues.
+
+### Goals
+
+* Make creating an IPv6 development environment as easy as creating an AWS
+  cluster, by just setting one additional option to turn it on.
+* Support OKD in reaching a milestone of supporting a single stack IPv6 control
+  plane (IPv6 only).
+
+### Non-Goals
+
+* Reaching full support for IPv6 on AWS.  While the work will get us closer,
+  that is not the primary goal here.  The primary goal is to help make it
+  easier for all teams to exercise IPv6.
+
+## Proposal
+
+To enable IPv6 in your AWS environment, a developer would set the following
+environment variable before running the installer:
+
+```bash
+    export OPENSHIFT_AWS_USE_IPV6=”true”
+```
+
+This would be off by default.
+
+### AWS Network Environment
+
+AWS does not support a single-stack (IPv6 only) environment, but it does
+support a dual-stack (IPv4 and IPv6) environment, so that’s what is enabled
+here.  This is a summary of the changes to the network environment:
+
+* The VPC has IPv6 enabled and a `/56` IPv6 CIDR will be allocated by AWS.
+* Each Subnet will have an IPv6 `/64` subnet allocated to it.
+* All IPv4 specific security group rules have corresponding IPv6 rules created.
+* AWS Network Load Balancers (NLBs) do not support IPv6, so external API access
+  is still over IPv4.  AWS does not have a TCP load balancer that supports
+  IPv6, other than classic load balancers with EC2-Classic, and not EC2-VPC.
+  AWS Application Load Balancers supposedly support IPv6, but that would
+  require doing HTTPS load balancing for the API instead of just TCP load
+  balancing, so we just use the IPv4 NLBs.  API access within the cluster is
+  still exercising IPv6.
+* IPv6 DNS records (AAAA) are created and the IPv4 (A) records are disabled,
+  except for the API.
+* IPv6 routing is configured.  Since all instances get global IPv6 addresses,
```

Comment:
from https://aws.amazon.com/premiumsupport/knowledge-center/configure-private-ipv6-subnet/

> If you want your IPv6 address to access the internet, but you don't want resources on the internet to initiate communication with your instance, you can use an egress-only internet gateway.
> -   Create an egress-only internet gateway in your Amazon VPC.
> -   Add a route to your route table that points all IPv6 traffic (::/0) or a specific range of IPv6 addresses to the egress-only internet gateway.

Can we make sure we have `resources on the internet to initiate communication with your instance` behavior required.

Reply:
Yeah, that's an interesting idea.  It looks like security groups are already set up to effectively provide the same result.  It seems like that's probably good enough for our purposes?

Reply:
>  It looks like security groups are already set up to effectively provide the same result

personally our users will expect we follow AWS best-practices. security groups are great but they can easily _fungible_. something more fundamental that provide the _private_ behavior would be much better appreciated by our users.

> It seems like that's probably good enough for our purposes?

If we can at least look into AWS recommendations and see if they are small enough to do now that would be best.. but if we should mention it as future work in the doc at least.

Reply:
Well, there's no users for this except as a dev/test platform, not to be supported for any customer use.  I get your point that an egress-only gateway is much more explicit and clear about the intended behavior though.  I'll change to that.  I'll make it clear in this doc that this minute it's a todo item, but I should be able to have it done Monday if not sometime sooner this weekend.